schema/scope_test.go

```diff
@@ -498,3 +498,54 @@ func TestApplyingExternalNamespaceToNonRefTypes(t *testing.T) {
 		})
 	}
 }
+
+func TestGetRootObject(t *testing.T) {
+	rootObject := schema.NewObjectSchema("a", map[string]*schema.PropertySchema{})
+	correctSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": rootObject,
+		},
+		RootValue: "a",
+	}
+	assert.Equals(t, correctSchema.RootObject(), rootObject)
+}
+
+func TestMismatchedRoot(t *testing.T) {
+	// This is a common user mistake: invalid RootValue in the scope.
+	// Panic when a Scope's RootValue is not a key in its ObjectsValue
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": schema.NewObjectSchema("a", map[string]*schema.PropertySchema{}),
+		},
+		RootValue: "wrong",
+	}
+	assert.PanicsContains(t, func() {
+		brokenSchema.RootObject()
+	}, "root object with ID \"wrong\" not found; available objects:\n\ta")
+}
+
+func TestNilRoot(t *testing.T) {
+	// This is just a bug case; nil object in the objects map.
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"a": nil,
+		},
+		RootValue: "a",
+	}
+	assert.PanicsContains(t, func() {
+		brokenSchema.RootObject()
+	}, "root object with ID \"a\" is nil")
+}
+
+func TestMismatchedRootID(t *testing.T) {
+	// This is a common user mistake: valid scope map key doesn't match the object's ID
+	brokenSchema := schema.ScopeSchema{
+		ObjectsValue: map[string]*schema.ObjectSchema{
+			"wrong": schema.NewObjectSchema("a", map[string]*schema.PropertySchema{}),
+		},
+		RootValue: "wrong",
+	}
```

Comment:
I feel like this is backwards:  the code should have `"a"` where it has `"wrong"` because those are _supposed_ to match each other, and it should have `"wrong"` where it has `"a"` because that's the mismatch case which the test is checking.

Reply:
There are two matching wrong values. It doesn't matter either way, but in this case we're testing the scope, so I'm clarifying that the object is correct, and they constructed the scope incorrectly. The map and root value are on the scope side, and the object is separate.

Reply:
"I'm so proud of you, son:  there were 150 people in that marching band, and everyone was out of step _except_ you!"  😆

If two of the three values _match_, I'd be strongly inclined to say that it's the _third_ one which is wrong!  🙃

The reason why it matters is that we can compare this test to the other three (ignoring the `nil` test...), and see that we've covered the success case and cases 1 and 2 and that this test is intended to cover case three by looking where `"wrong"` appears among the three values.  By putting it in two places in this test, it makes it hard to figure out how this test relates to the other two.

> they constructed the scope incorrectly

Um, who is the "they" in that assertion?  😆  What the test is simulating is the user's YAML file, and that file specifies three values which get reflected here, and _one_ of them is wrong.  😉